# Introduction to Recommender Systems

<p align="center">
    <img width="721" alt="cover-image" src="https://user-images.githubusercontent.com/49638680/204351915-373011d3-75ac-4e21-a6df-99cd1c552f2c.png">
</p>

---

# ALS and Matrix Factorisation

The __Alternating Least Squares__ algorithm is a popular technique used in recommender systems for collaborative filtering. 
As you might remember, collaborative filtering is a technique that uses the past behaviour of users to make recommendations for new items. 
The goal of the ALS algorithm is to factorise a user-item matrix into two matrices, one representing the users and the other representing the items, so that the dot product of the two matrices approximates the original matrix.

The ALS algorithm works by alternately fixing one of the matrices while optimising the other. 
Specifically, it alternates between fixing the user matrix and optimising the item matrix and fixing the item matrix and optimising the user matrix. This alternating process continues until a certain convergence criteria is met.

The optimisation process involves minimising the difference between the predicted ratings and the actual ratings in the user-item matrix. This is done by computing the _least squares_ solution for each matrix. The algorithm uses a regularization term to prevent overfitting and ensure that the solution is stable.

The ALS algorithm can handle large and sparse datasets by breaking down the computation into smaller sub-problems. This allows the algorithm to scale to millions of users and items. The algorithm can also handle implicit feedback data where the absence of a rating is not necessarily a negative signal.

A further advantage of the ALS algorithm is that it is computationally efficient and can be parallelized easily. This makes it suitable for large-scale recommender systems used by companies such as Amazon and Netflix. The algorithm is also robust to noise and can handle missing data.

However, one limitation of the ALS algorithm is that it assumes that the user-item matrix can be factorised into two matrices. This assumption may not hold in some cases, leading to poor recommendations. Additionally, the ALS algorithm may not perform well in situations where there are few ratings for each user or item.

In conclusion, the ALS algorithm is a widely used technique in collaborative filtering-based recommender systems. It is computationally efficient, scalable, and can handle large and sparse datasets. However, it may not perform well in all situations and its assumptions may not hold in some cases.

## A common problem 

In all the algorithms for recommender systems we have seen up to this point, we could find some common issues:

1. _Popularity bias_: recommendations were imbalanced towards the most "popular" items, the ones with a great number of ratings. In general, this refers to system recommends the movies with the most interactions without any personalisation.
2. _Cold start_: recommendations for users, items or both with a low number of ratings were not really accurate. In general, this refers to when movies added to the catalogue have either none or very little interactions while recommender rely on the movie’s interactions to make recommendations.
3. _Scalability issue_: if the underlying training database is too large, we would struggle to fit our model. In generale, this refers to the lack of ability to scale to much larger sets of data when more and more users and movies got added into our database.

All the three above are very typical challenges for collaborative filtering recommender. 

They arrive naturally along with the user-movie (or movie-user) interaction matrix where each entry records an interaction of a user $i$ and a movie $j$. In a real world setting, the vast majority of movies receive very few or even no ratings at all by users. We are looking at an extremely sparse matrix with more than $99\%$ of entries are missing values.

With such a sparse matrix, what ML algorithm can be trained and reliable to make inference? 

To find solutions to the question, we are effectively solving a data sparsity problem.

## Matrix Factorization

In collaborative filtering, matrix factorisation is the state-of-the-art solution for sparse data problem, although it has become widely known since [Netflix Prize Challenge](https://www.netflixprize.com/).

<p align="center">
    <img width="650" src="https://sigopt.com/wp-content/uploads/2018/10/collaborative_filtering.png">
</p>

Matrix factorisation is a family of mathematical operations for matrices in linear algebra. 

To be specific, a matrix factorisation is a decomposition of a matrix $A$ into a product of matrices $B, C$ of suitable dimensions.

$$ A \simeq B \times C \, .$$ 

In the case of collaborative filtering, matrix factorization algorithms work by decomposing the user-item interaction matrix into the product of two lower dimensionality rectangular matrices. 
One matrix can be seen as the user matrix where rows represent users and columns are latent factors. 
The other matrix is the item matrix where rows are latent factors and columns represent items.

### How does matrix factorisation solve our problems?

There are several aspects of matrix factorisation that can be exploited to solve the issues described above.

1. The matrix factorisation-based model learns the user-item interaction matrix factorisation by user and item representation matrices. This allows the model to predict better personalised movie ratings for users. (Addresses Popularity bias and cold start)

2. With matrix factorisation, less-known items can have the same rich latent representation of "popular" ones. This improves recommender's ability to suggest less-known movies. (Addresses Popularity bias)

3. Matrix factorisation algorithm is really efficient at prediction time, since it is just a matter of multiplying two matrices. (Addresses scalability issue)

4. The matrix factorisation algorithm is parallelisable. (Addresses scalability issue)


### Matrix factorisation working scheme

As in collaborative filtering, we have the problem of predicting ratings out of a very sparse matrix.

Schematically, given an user $j$ and an item $i$, we can denote the user rating for the item as $y_{ij}$.

The matrix factorisation algorithm core hypothesis is that $y_{ij}$ can be expressed as a matrix product:

$$ y_{ij} = \sum_{k} \mathcal{H}_{ik} \mathcal{H}_{kj} \, . $$

Or in a more general notation,

$$ Y = \mathcal{H}_m \times \mathcal{H}_m \, , $$

where $\mathcal{H}_u$ and $\mathcal{H}_m$ are the latent matrices of users and items respectively. These are the factors in __matrix factorisation__.

There is a tunable hyperparameter in the matrix factorisation above, the number of latent variables, _i.e._ the dimension of the matrices $\mathcal{H}$. 
The tuning can be done by __cross-validation__, which is a crucial procedure in machine learning.

__Latent factors__ are the features in the lower dimension latent space projected from user-item interaction matrix. 
The idea behind matrix factorisation is to use latent factors to represent user preferences or movie topics in a much lower dimension space. 
Matrix factorisation is one of very effective dimension reduction techniques in machine learning.

Now the question is how to find these matrices $\mathcal{H}_u$ and $\mathcal{H}_m$?

We are going to explore two different approaches: 

1. Singular Value Decomposition - for those that are interested in the formalities, a nice reference can be found at [this link](http://www.math.iit.edu/~fass/477577_Chapter_12.pdf).
2. Alternating Least Squares - again, for a much deeper explanation, one can look [here](http://stanford.edu/~rezab/classes/cme323/S15/notes/lec14.pdf).

#### Singular Value Decomposition

[Singular Value Decomposition](https://en.wikipedia.org/wiki/Singular_value_decomposition) or __SVD__, is a matrix factorisation algorithm that generalises the eigenvalues/eigenvector decomposition to rectangular matrices.

__SVD__ is similar to Principal Component Analysis (PCA), but more general. PCA assumes that input are square matrices, SVD does not have this assumption. 

The general formula for SVD is:

$$ M = U \Sigma V^t \, ,$$

where

* $M$ is the original matrix to decompose.
* $U$ is a left singular value matrix.
* $V$ is a right singular value matrix.
* $\Sigma$ is a diagonal matrix containing singular values.

Recall: 
 - left singular value matrices are matrices whose columns are eigenvectors of $MM^t$.
 - right singular value matrices are matrices whose columns are eigenvectors of $M^tM$.

<p align="center">
    <img src="https://miro.medium.com/v2/resize:fit:720/format:webp/1*mo8loFarEKeNeVX49205-g.png">
</p>

One can state that SVD is more general than PCA. 

Indeed, SVD can handle matrices with different number of columns and rows.
PCA formula is $ M=Q\Lambda Q^t$, which decomposes matrix into orthogonal matrix Q and diagonal matrix $\Lambda$. 
PCA can be simply interpreted as:

1. Change of the basis (from the standard basis to the eigenvector one).
2. Application of the transformation $\Lambda$, that does not change the direction of vectors (diagonal matrix).
3. Back to the standard basis (using the matrix $Q$).

SVD does similar things, but it does not return to same basis from which we started transformations. 

It could not do it because our original matrix $M$ is not a square matrix. 
The following picture shows how the change of basis and the transformations are related to SVD.


<p align="center">
    <img src="https://upload.wikimedia.org/wikipedia/commons/b/bb/Singular-Value-Decomposition.svg">
</p>

Analogously, one can interpret SVD decomposition as follows:

1. Change of the basis from standard basis to basis $V$ (using $V^t$). Note that in graph this is represented as a simple rotation.
2. Apply the transformation described by $\Sigma$. This is a scaling of vectors in this basis.
3. Change the basis from $V$ to $U$.



<p align="center">
    <img src="https://upload.wikimedia.org/wikipedia/commons/e/e9/Singular_value_decomposition.gif">
</p>

##### SVD calculations

How to calculate SVD decomposition? 
There are several ways to do so. The typical one is solving something similar to an eigenvalue decomposition.

The exact solution to this problem is beyond the scope of these lectures, hence we are going to focus on the numerical one.

The numpy and scipy libraries have svd methods implemented already, let's look at how to use them.

In [1]:
# Import libraries
import numpy as np
import pandas as pd
import ast

from tabulate import tabulate

import matplotlib.pyplot as plt

# Set plot parameters
plt.rcParams["figure.figsize"] = (20, 13)
%matplotlib inline
%config InlineBackend.figure_format = "retina"

In [2]:
# SVD by numpy

## Initialise a random matrix
a = np.random.randint(1, 100, size=(12, 10))

## Decompose the matrix by SVD
u, s, vh = np.linalg.svd(a, full_matrices=True)

print(u.shape, s.shape, vh.shape)

(12, 12) (10,) (10, 10)


In [3]:
print(tabulate(a))

--  --  --  --  --  --  --  --  --  --
16  54  12  54   2  99  39  37  41  26
12  63  76  57  13  41  27  84  69  57
76  22  66  82   7  88  27  58  81   6
36  66  29  24  81  88  33  97  86  61
50  46   8  73  50  11  74   6  65   5
 8   9  23  83   5  65  33  67   3  77
66   3  56  50  96  34  14  12  62  99
 8  74  64  57  37  74  62  97  44  91
14  85  96  98  62  80  39  41  57  55
31  17  38  44  97  58  42  51  25  19
21  54  18  83  19  22  36  21  42  87
62  46  12  97  82  35  16  37  28  17
--  --  --  --  --  --  --  --  --  --


In [4]:
# Recompose the diagonal matrix Sigma
sigma = s*np.eye(10)
sigma = np.vstack((sigma, np.zeros(10), np.zeros(10)))

print(tabulate(sigma))

-------  -------  -------  -------  -------  -------  -------  -------  -------  -------
542.509    0        0        0       0        0        0        0        0        0
  0      149.052    0        0       0        0        0        0        0        0
  0        0      121.196    0       0        0        0        0        0        0
  0        0        0      111.351   0        0        0        0        0        0
  0        0        0        0      90.8308   0        0        0        0        0
  0        0        0        0       0       87.7241   0        0        0        0
  0        0        0        0       0        0       75.1864   0        0        0
  0        0        0        0       0        0        0       58.7445   0        0
  0        0        0        0       0        0        0        0       50.7919   0
  0        0        0        0       0        0        0        0        0       16.3661
  0        0        0        0       0        0        0        0 

In [5]:
sigma.shape

(12, 10)

In [6]:
a_rebuilt = np.dot(np.dot(u,sigma), vh)
print(tabulate(a_rebuilt))

--  --  --  --  --  --  --  --  --  --
16  54  12  54   2  99  39  37  41  26
12  63  76  57  13  41  27  84  69  57
76  22  66  82   7  88  27  58  81   6
36  66  29  24  81  88  33  97  86  61
50  46   8  73  50  11  74   6  65   5
 8   9  23  83   5  65  33  67   3  77
66   3  56  50  96  34  14  12  62  99
 8  74  64  57  37  74  62  97  44  91
14  85  96  98  62  80  39  41  57  55
31  17  38  44  97  58  42  51  25  19
21  54  18  83  19  22  36  21  42  87
62  46  12  97  82  35  16  37  28  17
--  --  --  --  --  --  --  --  --  --


In [7]:
assert np.isclose(a, a_rebuilt).all(), "SVD error, matrix is not rebuilt correctly."

#### SVD in a recommender system

Let’s see how we can make use of the result from SVD to build a recommender system. 

Firstly, let’s download the dataset from [this link](https://drive.google.com/file/d/1wgXv14TyrRD5DH6ZfJqApFymQv7_vuho/view).

In [30]:
%%bash

wget -N --load-cookies /tmp/cookies.txt "https://drive.google.com/u/0/uc?id=1wgXv14TyrRD5DH6ZfJqApFymQv7_vuho&export=download&confirm=$(wget --quiet --save-cookies /tmp/cookies.txt --keep-session-cookies --no-check-certificate 'https://docs.google.com/uc?export=download&id=1wgXv14TyrRD5DH6ZfJqApFymQv7_vuho' -O- | sed -rn 's/.*confirm=([0-9A-Za-z_]+).*/\1\n/p')&id=1wgXv14TyrRD5DH6ZfJqApFymQv7_vuho" -O data_books.tar.gz && rm -rf /tmp/cookies.txt
if [ ! -d "../data/books" ]; then mkdir -p ../data/books; fi
tar -zxf data_books.tar.gz --directory ../data/books

for details.

--2023-03-13 18:09:03--  https://drive.google.com/u/0/uc?id=1wgXv14TyrRD5DH6ZfJqApFymQv7_vuho&export=download&confirm=t&id=1wgXv14TyrRD5DH6ZfJqApFymQv7_vuho
Resolving drive.google.com... 2a00:1450:4007:819::200e, 142.250.178.142
Connecting to drive.google.com|2a00:1450:4007:819::200e|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://drive.google.com/uc?id=1wgXv14TyrRD5DH6ZfJqApFymQv7_vuho&export=download&confirm=t&id=1wgXv14TyrRD5DH6ZfJqApFymQv7_vuho [following]
--2023-03-13 18:09:03--  https://drive.google.com/uc?id=1wgXv14TyrRD5DH6ZfJqApFymQv7_vuho&export=download&confirm=t&id=1wgXv14TyrRD5DH6ZfJqApFymQv7_vuho
Reusing existing connection to [drive.google.com]:443.
HTTP request sent, awaiting response... 303 See Other
Location: https://doc-0k-6o-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/41or4ju5n4m5si0e7tg1uikrgqla1ban/1678727325000/00230115444143604265/*/1wgXv14TyrRD5DH6ZfJqApFymQv7_vuho?e=download&uui

##### The dataset

This dataset is the “_Social Recommendation Data_” from “Recommender Systems and Personalization Datasets“. 
It contains the reviews given by users on books on [Librarything](librarything.com). 

What we are interested are the number of “stars” a user given to a book.

If we open up this tar file we will see a large file named “reviews.txt”. We can extract it, or read the included file on the fly. 
The first three lines of reviews file are shown below:

In [28]:
file_path = "../data/books/lthing_data/reviews.txt"

count = 0
with open(file_path) as f:
    for line in f:
        print(line)
        count += 1
        if count > 3:
            break

reviews[('8167221', 'dianaleez')] = {'comment': "If a novel can be written in crisp black and white images, with a disillusioned young hero (imagine here a sleepy-eyed young Robert Mitchum), and a cast of tragic, yet stock, characters, Joseph Kannons 'Stardust' hits its mark. Kannons homage to the film noir both rings with authenticity and provides a compelling read.\nBen Collier returns from post-war Germany to discover that his brother Danny has inexplicitly committed suicide, leaving behind a widow, the luscious and lethal Liesel. Why did Danny fall to his death? Did he jump? Was he pushed? Why does no one other than Ben seem to care?\n'Stardust' is a character study, recounting the lives of those displaced by Hitler who will never feel secure again. Should they return to Germany? And to which Germany, East or West? It is a romance, as Danny and Liesel are strongly drawn to each other (I found it impossible not to see Mitchum and Hedy Lamar facing each other through hazy cigarette s

We have a jsonline file, meaning that each line in reviews.txt is a record in the form of a json object.

We are going to extract the `user`, `work`, and `stars` fields of each record as long as there are no missing data among these three. Despite the name, the records are not well-formed JSON strings (most notably it uses single quote rather than double quote). Therefore, we cannot use json package from Python but we need to use _ast_ to decode such string.

In [31]:
reviews = []

cnt = 0
with open(file_path) as f:
    for i, line in enumerate(f):
        line = line.split("=")[-1]
        
        try:
            record = ast.literal_eval(line)
        except SyntaxError:
            continue
        except ValueError:
            continue

        try:
            if any(x not in record for x in ['user', 'work', 'stars']):
                continue
        except TypeError:
            continue
        
        reviews.append((record['user'], record['work'], record['stars']))

print(f"{len(reviews)} reviews retrieved")

1377339 reviews retrieved)


Now we should create a matrix of how different users rate each book. 
We can make use of the pandas library to help convert the data we collected into a table.

In [34]:
reviews = pd.DataFrame(reviews, columns=["user", "work", "stars"])
reviews.head()

user      work  stars
0  Elizabeth.Wong98     73960    4.5
1            rivkat     69413    3.0
2      suz.haugland   9523995    4.0
3        amoskovacs    368228    4.0
4            CandyH  11243828    4.0

As you can imagine, we do not want to use all the data, this would full our memory quite quickly. 

Here, we consider only those users who reviewed more than $50$ books and also those books who are reviewed by more than $50$ users. 

This way, we trimmed our dataset to less than $15\%$ of its original size.

In [35]:
# Look for the users who reviewed more than 50 books
usercount = reviews[["work","user"]].groupby("user").count()
usercount = usercount[usercount["work"] >= 50]
usercount.head()

work
user            
              84
-Eva-        600
06nwingert   370
1983mk        63
1dragones    193

In [36]:
# Look for the books who reviewed by more than 50 users
workcount = reviews[["work","user"]].groupby("work").count()
workcount = workcount[workcount["user"] >= 50]
workcount.head()


user
work          
10000      106
10001       53
1000167    185
10001797    53
10005525   132

In [38]:
# Filter only the popular books and active users
reviews = reviews[reviews["user"].isin(usercount.index) & reviews["work"].isin(workcount.index)]
reviews

user     work  stars
5             miyurose  9071901    2.0
8           Mamajeanne  1110874    5.0
11           funkendub     5852    5.0
12       notmyrealname  3620689    2.0
13           bluetyson     1472    5.0
...                ...      ...    ...
1377314          Jim53    30888    3.5
1377318      lucybrown     3253    3.5
1377319      ElizaJane  7874593    4.5
1377331      heidijane  2129329    4.0
1377337      tamaranth  4873693    4.5

[202508 rows x 3 columns]

As in other cases, converting this to a matrix it is a matter of using a pivot method in pandas.

In [40]:
reviewmatrix = reviews.pivot(index="user", columns="work", values="stars").fillna(0)
reviewmatrix

work        10000  10001  1000167  10001797  10005525  10007394  10007399  \
user                                                                        
              0.0    0.0      0.0       0.0       0.0       0.0       0.0   
-Eva-         0.0    0.0      0.0       0.0       0.0       0.0       0.0   
06nwingert    0.0    0.0      0.0       0.0       0.0       0.0       0.0   
1983mk        0.0    0.0      0.0       0.0       0.0       0.0       0.0   
1dragones     5.0    4.0      0.0       0.0       0.0       0.0       0.0   
...           ...    ...      ...       ...       ...       ...       ...   
zjakkelien    0.0    0.0      0.0       0.0       0.0       0.0       0.0   
zmagic69      0.0    0.0      0.0       0.0       0.0       0.0       0.0   
zquilts       0.0    0.0      0.0       0.0       0.0       0.0       0.0   
zwaantje      0.0    0.0      0.0       0.0       0.0       0.0       0.0   
zzshupinga    0.0    0.0      0.0       0.0       0.0       0.0       0.0   

work        10009  10012725  10012975  ...  9978  9979582  9984  9986454  \
user                                   ...                                 
              0.0       0.0       0.0  ...   0.0      0.0   0.0      0.0   
-Eva-         0.0       0.0       0.0  ...   0.0      3.5   0.0      0.0   
06nwingert    0.0       0.0       0.0  ...   0.0      0.0   0.0      0.0   
1983mk        0.0       0.0       0.0  ...   0.0      0.0   0.0      0.0   
1dragones     0.0       0.0       0.0  ...   0.0      0.0   0.0      0.0   
...           ...       ...       ...  ...   ...      ...   ...      ...   
zjakkelien    0.0       0.0       0.0  ...   0.0      0.0   0.0      0.0   
zmagic69      0.0       0.0       0.0  ...   0.0      0.0   0.0      0.0   
zquilts       0.0       0.0       0.0  ...   0.0      0.0   0.0      0.0   
zwaantje      0.0       0.0       0.0  ...   0.0      0.0   0.0      0.0   
zzshupinga    0.0       0.0       0.0  ...   0.0      0.0   0.0      0.0   

work        9989632  9989655  9989664  9993  9997232  9998  
user                                                        
                0.0      0.0      0.0   0.0      0.0   0.0  
-Eva-           0.0      0.0      0.0   0.0      0.0   0.0  
06nwingert      0.0      0.0      0.0   0.0      0.0   0.0  
1983mk          0.0      0.0      0.0   0.0      0.0   0.0  
1dragones       0.0      4.0      0.0   0.0      0.0   0.0  
...             ...      ...      ...   ...      ...   ...  
zjakkelien      0.0      0.0      0.0   0.0      0.0   0.0  
zmagic69        0.0      0.0      0.0   0.0      0.0   0.0  
zquilts         0.0      0.0      0.0   0.0      0.0   0.0  
zwaantje        0.0      0.0      0.0   0.0      0.0   0.0  
zzshupinga      0.0      0.0      0.0   0.0      0.0   0.0  

[5563 rows x 2872 columns]

Here, we represented $5563$ users and $2872$ books in a matrix. 

Hence, we can apply the SVD algorithm. 

In [41]:
# Careful, this might take a while
matrix = reviewmatrix.values
u, s, vh = np.linalg.svd(matrix, full_matrices=False)

We have seen that by default, the `svd` returns a full singular value decomposition. 

We choose a reduced version so we can use smaller matrices to save memory. 
The columns of vh correspond to the books. 

We can exploit vector space structure to find which book are most similar to the one we are looking at, by looking at the distance of elements. 
Since we are dealing with a great number of elements, it is more efficient to look at the cosine distance.

In [42]:
def cosine_similarity(v: np.ndarray, u: np.ndarray) -> float:
    """Calculate the cosine distance between two arrays.

    Parameters
    ----------
    v : np.ndarray
        An array
    u : np.ndarray
        An array

    Returns
    -------
    float
        the cosine distance between the two arrays.
    """
    return (np.dot(v, u))/(np.linalg.norm(v) * np.linalg.norm(u))

We are ready to assign a score, based on distance to all the elements in the database.

In [44]:
item_id = 0

highest_similarity = -np.inf
highest_sim_col = -1
for col in range(1, vh.shape[1]):
    similarity = cosine_similarity(vh[:,item_id], vh[:,col])
    if similarity > highest_similarity:
        highest_similarity = similarity
        highest_sim_col = col
 
print(f"Vector {highest_sim_col} is most similar to vector {item_id}")

Vector 2623 is most similar to vector 0


And in the above example, we try to find the book that is best match to to first column. 

The result is $2623$.

In a recommendation system, when a user picked a book, we may show her a few other books that are similar to the one she picked based on the cosine distance as calculated above.

Depends on the dataset, we may use truncated SVD to reduce the dimension of matrix vh. In essence, this means we are removing several rows on vh that the corresponding singular values in s are small, before we use it to compute the similarity. This would likely make the prediction more accurate as those less important features of a book are removed from consideration.

Note that, in the decomposition $M = U \Sigma V^t $ we know the rows of $U$ are the users and columns of $V^t$ are items, we cannot identify what are the meanings of the columns of $U$ nor rows of $V^t$ (an equivalently, that of $\Sigma$). 
We know they could be genres, for example, that provide some underlying connections between the users and the books but we cannot be sure what exactly are they. 
However, this does not stop us from using them as features in our recommendation system.